### PR TITLE
Add slack notifications to pipeline builds

### PR DESCRIPTION
**Pull request by AdamBrousseau**
- If a downstream job fails, send a slack
- Add slack_channel to variables file
- Slack message contains Pipeline build name, number, url
  as well as downstream job that failed name, number, url

[skip ci]

Issue #975

Signed-off-by: Adam Brousseau <adam.brousseau88@gmail.com>

### CODE REVIEW

**Comment by AdamBrousseau** [comment]
See [Slack](https://openj9.slack.com/archives/C8QU16TDW/p1520283781000456) for some examples

File: buildenv/jenkins/common/pipeline-functions [hunk 1]
```diff
@@ -78,15 +78,44 @@ def get_sha(REPO, BRANCH) {
...
     stage ('Build') {
-        JOB = build job: BUILD_JOB_NAME, parameters: [string(name: 'OPENJDK_REPO', value: OPENJDK_REPO), string(name: 'OPENJDK_BRANCH', value: OPENJDK_BRANCH), string(name: 'OPENJDK_SHA', value: OPENJDK_SHA), string(name: 'OPENJ9_REPO', value: OPENJ9_REPO), string(name: 'OPENJ9_BRANCH', value: OPENJ9_BRANCH), string(name: 'OPENJ9_SHA', value: OPENJ9_SHA), string(name: 'OMR_REPO', value: OMR_REPO), string(name: 'OMR_BRANCH', value: OMR_BRANCH), string(name: 'OMR_SHA', value: OMR_SHA), string(name: 'VARIABLE_FILE', value: VARIABLE_FILE)]
-        return JOB
+        return build_with_slack(BUILD_JOB_NAME, [string(name: 'OPENJDK_REPO', value: OPENJDK_REPO), string(name: 'OPENJDK_BRANCH', value: OPENJDK_BRANCH), string(name: 'OPENJDK_SHA', value: OPENJDK_SHA), string(name: 'OPENJ9_REPO', value: OPENJ9_REPO), string(name: 'OPENJ9_BRANCH', value: OPENJ9_BRANCH), string(name: 'OPENJ9_SHA', value: OPENJ9_SHA), string(name: 'OMR_REPO', value: OMR_REPO), string(name: 'OMR_BRANCH', value: OMR_BRANCH), string(name: 'OMR_SHA', value: OMR_SHA), string(name: 'VARIABLE_FILE', value: VARIABLE_FILE)])
```
Comment (DanHeidinga) [inline buildenv/jenkins/common/pipeline-functions]: Can these be formatted in a more readable way?  Maybe like:
```
return build_with_slack(BUILD_JOB_NAME, 
    [string(name: 'OPENJDK_REPO', value: OPENJDK_REPO), 
    string(name: 'OPENJDK_BRANCH', value: OPENJDK_BRANCH), 
    string(name: 'OPENJDK_SHA', value: OPENJDK_SHA), 
    string(name: 'OPENJ9_REPO', value: OPENJ9_REPO),    
    string(name: 'OPENJ9_BRANCH', value: OPENJ9_BRANCH), 
    string(name: 'OPENJ9_SHA', value: OPENJ9_SHA), 
    string(name: 'OMR_REPO', value: OMR_REPO), 
    string(name: 'OMR_BRANCH', value: OMR_BRANCH), 
    string(name: 'OMR_SHA', value: OMR_SHA), 
    string(name: 'VARIABLE_FILE', value: VARIABLE_FILE)])
```

File: buildenv/jenkins/common/pipeline-functions [hunk 2]
```diff
@@ -78,15 +78,44 @@ def get_sha(REPO, BRANCH) {
...
     }
 }
 
 def build_with_one_upstream(JOB_NAME, UPSTREAM_JOB_NAME, UPSTREAM_JOB_NUMBER, VARIABLE_FILE) {
     stage ("${JOB_NAME}") {
-        JOB = build job: JOB_NAME, parameters: [string(name: 'UPSTREAM_JOB_NAME', value: UPSTREAM_JOB_NAME), string(name: 'UPSTREAM_JOB_NUMBER', value: "${UPSTREAM_JOB_NUMBER}"), string(name: 'VARIABLE_FILE', value: VARIABLE_FILE)]
-        return JOB
+        return build_with_slack(JOB_NAME, [string(name: 'UPSTREAM_JOB_NAME', value: UPSTREAM_JOB_NAME), string(name: 'UPSTREAM_JOB_NUMBER', value: "${UPSTREAM_JOB_NUMBER}"), string(name: 'VARIABLE_FILE', value: VARIABLE_FILE)])
```
Comment (DanHeidinga) [inline buildenv/jenkins/common/pipeline-functions]: Same question about formatting

File: buildenv/jenkins/common/pipeline-functions [hunk 3]
```diff
@@ -78,15 +78,44 @@ def get_sha(REPO, BRANCH) {
 
 def build(BUILD_JOB_NAME, OPENJDK_REPO, OPENJDK_BRANCH, OPENJDK_SHA, OPENJ9_REPO, OPENJ9_BRANCH, OPENJ9_SHA, OMR_REPO, OMR_BRANCH, OMR_SHA, VARIABLE_FILE) {
```
Comment (DanHeidinga) [inline buildenv/jenkins/common/pipeline-functions]: Where is `build` called from?  Only by pipelines?  I'm trying to trace who may end up calling `build_with_slack` but the `build` is hard to grep :)

File: buildenv/jenkins/common/pipeline-functions (hunk 1 above)
Comment (AdamBrousseau) [reply buildenv/jenkins/common/pipeline-functions]: Shouldn't be a problem to change that

File: buildenv/jenkins/common/pipeline-functions (hunk 3 above)
Comment (AdamBrousseau) [reply buildenv/jenkins/common/pipeline-functions]: yes, only pipelines (i.e nightly builds and eventually OMR-Acceptance)
https://github.com/eclipse/openj9/blob/master/buildenv/jenkins/common/pipeline-functions#L98
https://github.com/eclipse/openj9/blob/master/buildenv/jenkins/common/pipeline-functions#L115

**Comment by AdamBrousseau** [comment]
Updated formatting, [still works](https://ci.eclipse.org/openj9/job/Adam_Pipeline-Build-Test-JDK8-linux_390-64_cmprssptrs/46/console)

**Comment by DanHeidinga** [comment]
Not running a PR build as this functionality is only tested by the nightlies.

**Pull request merged**